src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -169,15 +169,14 @@ protected function refreshUser(TokenInterface $token)
 
             try {
                 $refreshedUser = $provider->refreshUser($user);
-                $token->setUser($refreshedUser);
+                $newToken = clone $token;
+                $newToken->setUser($refreshedUser);
 
                 // tokens can be deauthenticated if the user has been changed.
-                if (!$token->isAuthenticated()) {
-                    if (null !== $this->logger) {
-                        $this->logger->debug('Token was deauthenticated after trying to refresh it.', array('username' => $refreshedUser->getUsername(), 'provider' => \get_class($provider)));
-                    }
+                if (!$newToken->isAuthenticated()) {
+                    $deauthenticationContext = array('username' => $refreshedUser->getUsername(), 'provider' => \get_class($provider));
 
-                    return null;
+                    continue;
                 }
```

Comment:
If `$newToken->isAuthenticated()` is true then `$token` is outdated at this stage, right? I think we miss a call to `$token->setUser($refreshedUser);`

Reply:
That's what I did at the begining but if `$newToken->isAuthenticated()` is true then that means that the user stored in the token and `$refreshedUser` are the same, no ? So I wonder if it is usefull ?

Reply:
I think it can be, I mean: you may want your user to keep being authenticated while some of its properties have changed, getting these properties up to date in the token storage is probably good. Deauthentication can be based on custom criteria (if the user implements EquatableInterface), one could allow a refreshed user to have e.g. new roles without deauthenticating, in this case having the new roles in the stored token would be desired I suppose

Reply:
You are right ! Thanks for the details. I made the changes.